File: examples/localcowork/mcp-servers/_shared/ts/mcp-base.ts
/**
 * MCP Server Base Classes — TypeScript
 *
 * Shared foundation for all TypeScript MCP servers in LocalCowork.
 * Implements JSON-RPC 2.0 over stdio transport and tool registration.
 *
 * Usage:
 *   import { MCPServer, MCPTool, MCPResult, MCPError } from '../../_shared/ts/mcp-base';
 */

import { ZodSchema } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

// ─── Types ──────────────────────────────────────────────────────────────────

/** Result returned by a tool execution */
export interface MCPResult<T = unknown> {
  success: boolean;
  data: T;
}

/** Structured error for MCP tool failures */
export class MCPError extends Error {
  code: number;

  constructor(code: number, message: string) {
    super(message);
    this.code = code;
    this.name = 'MCPError';
  }
}

/** Standard MCP error codes */
export const ErrorCodes = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603,
  // Custom codes for LocalCowork
  SANDBOX_VIOLATION: -32001,
  CONFIRMATION_REQUIRED: -32002,
  FILE_NOT_FOUND: -32003,
  PERMISSION_DENIED: -32004,
} as const;

/** Tool definition interface — every tool implements this */
export interface MCPTool<TParams = unknown> {
  /** Fully qualified tool name: server.tool_name */
  name: string;

  /** Human-readable description for the LLM */
  description: string;

  /** Zod schema for parameter validation */
  paramsSchema: ZodSchema<TParams>;

  /** Whether the Agent Core must get user confirmation before executing */
  confirmationRequired: boolean;

  /** Whether this action can be undone via the undo stack */
  undoSupported: boolean;

  /** Execute the tool with validated params */
  execute(params: TParams): Promise<MCPResult>;
}

// ─── JSON-RPC Types ─────────────────────────────────────────────────────────

interface JsonRpcRequest {
  jsonrpc: '2.0';
  id: string | number;
  method: string;
  params?: Record<string, unknown>;
}

interface JsonRpcResponse {
  jsonrpc: '2.0';
  id: string | number;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
}

// ─── MCPServer ──────────────────────────────────────────────────────────────

export interface MCPServerConfig {
  name: string;
  version: string;
  tools: MCPTool[];
}

/**
 * Base MCP Server class.
 *
 * Registers tools, handles JSON-RPC over stdio, validates params,
 * and dispatches tool calls.
 *
 * Usage:
 *   const server = new MCPServer({ name: 'filesystem', version: '1.0.0', tools: [...] });
 *   server.start();
 */
export class MCPServer {
  private name: string;
  private version: string;
  private tools: Map<string, MCPTool>;

  constructor(config: MCPServerConfig) {
    this.name = config.name;
    this.version = config.version;
    this.tools = new Map();

    for (const tool of config.tools) {
      this.tools.set(tool.name, tool);
    }
  }

  /** Start the JSON-RPC listener on stdio */
  start(): void {
    process.stdin.setEncoding('utf-8');

    let buffer = '';

    process.stdin.on('data', (chunk: string) => {
      buffer += chunk;

      // Try to parse complete JSON-RPC messages (newline-delimited)
      const lines = buffer.split('\n');
      buffer = lines.pop() ?? '';

      for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed.length === 0) continue;

        try {
          const request = JSON.parse(trimmed) as JsonRpcRequest;
          this.handleRequest(request).then((response) => {
            if (response) {
              process.stdout.write(JSON.stringify(response) + '\n');
            }
          });
        } catch {
          const error: JsonRpcResponse = {
            jsonrpc: '2.0',
            id: 0,
            error: { code: ErrorCodes.PARSE_ERROR, message: 'Invalid JSON' },
          };
          process.stdout.write(JSON.stringify(error) + '\n');
        }
      }
    });

    process.stdin.on('end', () => {
      process.exit(0);
    });

    // No proactive init message — the Rust client sends an `initialize`
    // request and we respond to it via handleRequest().
  }

  /**
   * Build the tool definitions array for the initialize/tools-list responses.
   *
   * Converts each registered tool's zod schema to JSON Schema using the
   * zod-to-json-schema library, producing full property/required/description
   * metadata that the LLM needs to call tools correctly.
   */
  private buildToolDefinitions(): Record<string, unknown>[] {
    return Array.from(this.tools.values()).map((tool) => ({
      name: tool.name,
      description: tool.description,
      inputSchema: zodToJsonSchema(tool.paramsSchema, {
        target: 'openApi3',
        $refStrategy: 'none',
      }),
      metadata: {
        confirmationRequired: tool.confirmationRequired,
        undoSupported: tool.undoSupported,
      },
    }));
  }

  /** Handle an incoming JSON-RPC request */
  private async handleRequest(request: JsonRpcRequest): Promise<JsonRpcResponse | null> {
    // MCP initialization handshake — Rust client sends this on server startup
    if (request.method === 'initialize') {
      return {
        jsonrpc: '2.0',
        id: request.id,
        result: {
          serverInfo: { name: this.name, version: this.version },
          tools: this.buildToolDefinitions(),
        },
      };
    }

    if (request.method === 'tools/call') {
      return this.handleToolCall(request);
    }

    if (request.method === 'tools/list') {
      return this.handleToolList(request);
    }

    if (request.method === 'ping') {
      return { jsonrpc: '2.0', id: request.id, result: { status: 'ok' } };
    }

    return {
      jsonrpc: '2.0',
      id: request.id,
      error: { code: ErrorCodes.METHOD_NOT_FOUND, message: `Unknown method: ${request.method}` },
    };
  }

  /** Handle a tools/call request */
  private async handleToolCall(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    const { name, arguments: args } = request.params as { name: string; arguments: unknown };

    const tool = this.tools.get(name);
    if (!tool) {
      return {
        jsonrpc: '2.0',
        id: request.id,
        error: { code: ErrorCodes.METHOD_NOT_FOUND, message: `Unknown tool: ${name}` },
      };
    }

    // Validate params
    const parseResult = tool.paramsSchema.safeParse(args);
    if (!parseResult.success) {
      return {
        jsonrpc: '2.0',
        id: request.id,
        error: {
          code: ErrorCodes.INVALID_PARAMS,
          message: `Invalid parameters: ${parseResult.error.message}`,
        },
      };
    }

    // Execute tool
    try {
      const result = await tool.execute(parseResult.data);
      return {
        jsonrpc: '2.0',
        id: request.id,
        result: {
          content: [{ type: 'text', text: JSON.stringify(result.data) }],
        },
      };
    } catch (err) {
      if (err instanceof MCPError) {
        return {
          jsonrpc: '2.0',
          id: request.id,
          error: { code: err.code, message: err.message },
        };
      }
      return {
        jsonrpc: '2.0',
        id: request.id,
        error: {
          code: ErrorCodes.INTERNAL_ERROR,
          message: `Internal error: ${err instanceof Error ? err.message : 'Unknown'}`,
        },
      };
    }
  }

  /** Handle a tools/list request */
  private handleToolList(request: JsonRpcRequest): JsonRpcResponse {
    return { jsonrpc: '2.0', id: request.id, result: { tools: this.buildToolDefinitions() } };
  }
}

File: examples/localcowork/mcp-servers/_shared/py/mcp_base.py
"""
MCP Server Base Classes — Python

Shared foundation for all Python MCP servers in LocalCowork.
Implements JSON-RPC 2.0 over stdio transport and tool registration.

Usage:
    from mcp_base import MCPServer, MCPTool, MCPResult, MCPError
"""

from __future__ import annotations

import asyncio
import json
import sys
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Generic, TypeVar

from pydantic import BaseModel, ValidationError

# ─── Type Variables ──────────────────────────────────────────────────────────

TParams = TypeVar("TParams", bound=BaseModel)
TResult = TypeVar("TResult", bound=BaseModel)

# ─── Result & Error Types ────────────────────────────────────────────────────


@dataclass
class MCPResult(Generic[TResult]):
    """Result returned by a tool execution."""

    success: bool
    data: TResult | None = None


class MCPError(Exception):
    """Structured error for MCP tool failures."""

    def __init__(self, code: int, message: str) -> None:
        super().__init__(message)
        self.code = code


class ErrorCodes:
    """Standard MCP error codes."""

    PARSE_ERROR = -32700
    INVALID_REQUEST = -32600
    METHOD_NOT_FOUND = -32601
    INVALID_PARAMS = -32602
    INTERNAL_ERROR = -32603
    # Custom codes for LocalCowork
    SANDBOX_VIOLATION = -32001
    CONFIRMATION_REQUIRED = -32002
    FILE_NOT_FOUND = -32003
    PERMISSION_DENIED = -32004


# ─── Tool Base Class ─────────────────────────────────────────────────────────


class MCPTool(ABC, Generic[TParams, TResult]):
    """
    Abstract base class for MCP tools.

    Every tool must define:
    - name: fully qualified (e.g., 'ocr.extract_text_from_image')
    - description: for the LLM
    - Params type: pydantic BaseModel for input validation
    - Result type: pydantic BaseModel for output structure
    - confirmation_required: whether Agent Core needs user confirmation
    - undo_supported: whether this action can be reversed
    - execute(): the implementation
    """

    name: str = ""
    description: str = ""
    confirmation_required: bool = False
    undo_supported: bool = False

    @abstractmethod
    async def execute(self, params: TParams) -> MCPResult[TResult]:
        """Execute the tool with validated parameters."""
        ...

    def get_params_model(self) -> type[BaseModel]:
        """Get the Pydantic model class for params validation."""
        # Extract from Generic type args
        for base in type(self).__orig_bases__:  # type: ignore[attr-defined]
            if hasattr(base, "__args__") and len(base.__args__) >= 1:
                return base.__args__[0]
        raise TypeError(f"Tool {self.name} must specify Generic params type")

    def get_input_schema(self) -> dict[str, Any]:
        """Generate JSON Schema from the Pydantic params model."""
        model = self.get_params_model()
        return model.model_json_schema()

    def to_definition(self) -> dict[str, Any]:
        """Generate the MCP tool definition for the Rust MCP client."""
        return {
            "name": self.name,
            "description": self.description,
            "params_schema": self.get_input_schema(),
            "confirmation_required": self.confirmation_required,
            "undo_supported": self.undo_supported,
        }


# ─── MCP Server ──────────────────────────────────────────────────────────────


class MCPServer:
    """
    Base MCP Server.

    Registers tools, handles JSON-RPC over stdio, validates params,
    and dispatches tool calls.

    Usage:
        server = MCPServer(
            name="ocr",
            version="1.0.0",
            tools=[ExtractTextFromImage(), ExtractTextFromPdf()],
        )
        server.start()
    """

    def __init__(self, name: str, version: str, tools: list[MCPTool[Any, Any]]) -> None:
        self.name = name
        self.version = version
        self.tools: dict[str, MCPTool[Any, Any]] = {}

        for tool in tools:
            self.tools[tool.name] = tool

    def start(self) -> None:
        """Start the JSON-RPC listener on stdio (blocking)."""
        asyncio.run(self._run())

    async def _run(self) -> None:
        """Main event loop: read stdin, dispatch, write stdout."""
        reader = asyncio.StreamReader()
        protocol = asyncio.StreamReaderProtocol(reader)
        await asyncio.get_event_loop().connect_read_pipe(lambda: protocol, sys.stdin)

        while True:
            line = await reader.readline()
            if not line:
                break  # stdin closed

            line_str = line.decode("utf-8").strip()
            if not line_str:
                continue

            try:
                request = json.loads(line_str)
                response = await self._handle_request(request)
                if response:
                    sys.stdout.write(json.dumps(response) + "\n")
                    sys.stdout.flush()
            except json.JSONDecodeError:
                error_resp = {
                    "jsonrpc": "2.0",
                    "id": 0,
                    "error": {"code": ErrorCodes.PARSE_ERROR, "message": "Invalid JSON"},
                }
                sys.stdout.write(json.dumps(error_resp) + "\n")
                sys.stdout.flush()

    def _build_init_result(self) -> dict[str, Any]:
        """Build the initialization result payload (tool manifest)."""
        tool_defs = [tool.to_definition() for tool in self.tools.values()]
        return {
            "server_info": {"name": self.name, "version": self.version},
            "tools": tool_defs,
            "capabilities": {},
        }

    async def _handle_request(self, request: dict[str, Any]) -> dict[str, Any] | None:
        """Dispatch a JSON-RPC request."""
        method = request.get("method", "")
        request_id = request.get("id", 0)

        if method == "initialize":
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "result": self._build_init_result(),
            }

        if method == "tools/call":
            return await self._handle_tool_call(request)

        if method == "tools/list":
            return self._handle_tool_list(request)

        if method == "ping":
            return {"jsonrpc": "2.0", "id": request_id, "result": {"status": "ok"}}

        return {
            "jsonrpc": "2.0",
            "id": request_id,
            "error": {
                "code": ErrorCodes.METHOD_NOT_FOUND,
                "message": f"Unknown method: {method}",
            },
        }

    async def _handle_tool_call(self, request: dict[str, Any]) -> dict[str, Any]:
        """Handle a tools/call request."""
        request_id = request.get("id", 0)
        params = request.get("params", {})
        tool_name = params.get("name", "")
        arguments = params.get("arguments", {})

        tool = self.tools.get(tool_name)
        if not tool:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {
                    "code": ErrorCodes.METHOD_NOT_FOUND,
                    "message": f"Unknown tool: {tool_name}",
                },
            }

        # Validate params
        try:
            params_model = tool.get_params_model()
            validated_params = params_model(**arguments)
        except ValidationError as e:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {
                    "code": ErrorCodes.INVALID_PARAMS,
                    "message": f"Invalid parameters: {e}",
                },
            }

        # Execute tool
        try:
            result = await tool.execute(validated_params)
            result_data = result.data
            if isinstance(result_data, BaseModel):
                result_data = result_data.model_dump()

            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "result": {
                    "content": [{"type": "text", "text": json.dumps(result_data)}],
                },
            }
        except MCPError as e:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {"code": e.code, "message": str(e)},
            }
        except Exception as e:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {
                    "code": ErrorCodes.INTERNAL_ERROR,
                    "message": f"Internal error: {e!s}",
                },
            }

    def _handle_tool_list(self, request: dict[str, Any]) -> dict[str, Any]:
        """Handle a tools/list request."""
        tool_defs = [tool.to_definition() for tool in self.tools.values()]
        return {
            "jsonrpc": "2.0",
            "id": request.get("id", 0),
            "result": {"tools": tool_defs},
        }
